### PR TITLE
[ML] Fix feature importance for binary classification

### LINKED ISSUE #1489: [ML] Incorrect output of feature importances for imbalanced classes
In `CDataFrameTrainBoostedTreeClassifierRunner::writeOneRow` we have
```c++
double importance{(j == predictedClassId)
                                                      ? shap[i](0)
                                                      : -shap[i](0)};
```

This is incorrect when the predicted class was selected on merits of class score and not class probability. This leads to feature importances summed up to a negative of the correct value.

### PR DESCRIPTION
**Pull request by valeriy42**
The reporting of feature importance for binary classification was made under (erroneous) assumption that the positive shap values support the predicted class. In fact, the positive shap values support the class 1 (the class that was read as second when were were reading the data frame). This PR fixes this. Since the new result format will be only available in 7.10, I mark this PR as `non-issue`.

Additionally, I tightened up the unit test. Instead of simply verifying that the bias term has small variance, I am explicitly computing the baseline and asserting that the local approximation with shap values is close to the predicted value. 

Closes #1489

### CODE REVIEW

File: lib/api/unittest/CDataFrameAnalyzerFeatureImportanceTest.cc [hunk 1]
```diff
@@ -614,8 +614,28 @@ BOOST_FIXTURE_TEST_CASE(testClassificationFeatureImportanceAllShap, SFixture) {
...
+            double c3{readShapValue(result, "c3", targetPrediction)};
+            double c4{readShapValue(result, "c4", targetPrediction)};
+            double predictionProbability{
+                result["row_results"]["results"]["ml"]["prediction_probability"].GetDouble()};
+            double logOdds = std::log(predictionProbability /
+                                      (1.0 - predictionProbability + 1e-10));
+            double cSum{c1 + c2 + c3 + c4};
+            // if predicted class is "foo", we need to flip the baseline
+            double baseline{(targetPrediction == "bar") ? baselineCommon : -baselineCommon};
+            BOOST_REQUIRE_SMALL(logOdds - (baseline + cSum), 1e-3);
```
Comment (tveasey) [inline lib/api/unittest/CDataFrameAnalyzerFeatureImportanceTest.cc:637]: 👍

**Review by tveasey — COMMENTED** [review]
Overall change looks good and good idea to tighten up the test assertion. My main comment is let's take the opportunity to fold the logic related to binary classification into `CTreeShapFeatureImportance`. This would clean up the result writing and improve encapsulation IMO.

File: lib/api/CDataFrameTrainBoostedTreeClassifierRunner.cc [hunk 2]
```diff
@@ -185,14 +185,15 @@ void CDataFrameTrainBoostedTreeClassifierRunner::writeOneRow(
...
                                 writer.StartObject();
                                 writer.Key(CLASS_NAME_FIELD_NAME);
                                 writePredictedCategoryValue(classValues[j], writer);
                                 writer.Key(IMPORTANCE_FIELD_NAME);
-                                writer.Double(importance);
+                                if (j == 1) {
+                                    writer.Double(shap[i](0));
+                                } else {
+                                    writer.Double(-shap[i](0));
+                                }
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/api/CDataFrameTrainBoostedTreeClassifierRunner.cc]: This logic is related to how we are formatting the shap values, not how they are computed. Therefore, I put this logic here. `CTreeShapFeatureImportance` lives in `ml::maths` doesn't really know if we are doing regression or classification.

File: lib/api/unittest/CDataFrameAnalyzerFeatureImportanceTest.cc [hunk 3]
```diff
@@ -614,8 +614,28 @@ BOOST_FIXTURE_TEST_CASE(testClassificationFeatureImportanceAllShap, SFixture) {
...
+            std::string targetPrediction{
+                result["row_results"]["results"]["ml"]["target_prediction"].GetString()};
+            double c1{readShapValue(result, "c1", targetPrediction)};
+            double c2{readShapValue(result, "c2", targetPrediction)};
+            double c3{readShapValue(result, "c3", targetPrediction)};
+            double c4{readShapValue(result, "c4", targetPrediction)};
+            double predictionProbability{
+                result["row_results"]["results"]["ml"]["prediction_probability"].GetDouble()};
+            double logOdds = std::log(predictionProbability /
+                                      (1.0 - predictionProbability + 1e-10));
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/api/unittest/CDataFrameAnalyzerFeatureImportanceTest.cc]: Fixed

**Comment by valeriy42** [comment]
@tveasey thank you for your comments. As we discussed, I also extended the multi-class classification part to allow for simple check that local approximation stacks-up to the prediction probability. I left the logic in the ClassifierRunner, since IMO it is related to how we are formatting the results and not how we are computing the shap values. If you have a bad feeling about it, I will try to move the logic somewhere else.

Since the feature importance values are re-scaled now, they don't sum-up to the total feature importance values. I commented out the relevant part of the unit test and left a to do to discuss/figure out a meaningfull way to compute scaled total feature importance in a follow-up PR.

Please, let me know what you think.

**Review by tveasey — APPROVED** [review]
> I left the logic in the ClassifierRunner, since IMO it is related to how we are formatting the results and not how we are computing the shap values. If you have a bad feeling about it, I will try to move the logic somewhere else.

I think on reflection what I'd really like to do is to unify the way we calculate binary feature importances and multi-class importances so we just have vector of the two class shap values. This need not hold up this fix so I'll go ahead and approve.

**Pull request merged**